### PR TITLE
Use Symfony contracts

### PR DESCRIPTION
**Pull request by phansys**
## Subject

Use Symfony contracts instead of deprecated classes.

I am targeting this branch, because these changes are fully BC.

## Changelog

```markdown

### Changed
- Replaced the usage of classes deprecated at `symfony/*:4.3.0` by its counterparts from `Symfony\Contracts\` namespace.
```

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -34,6 +34,7 @@
         "symfony/console": "^2.8 || ^3.2 || ^4.0",
         "symfony/dependency-injection": "^2.8 || ^3.2 || ^4.0",
         "symfony/event-dispatcher": "^2.8 || ^3.2 || ^4.0",
```
Comment (core23) [inline composer.json]: Wo we even need this dependency?

Comment (phansys) [reply composer.json]: Yes, there are several usages of the classes provided by this dependency:
https://github.com/sonata-project/SonataAdminBundle/blob/a2524e7a6f03de2f8b8e84203088f6e57b209a30/src/Event/AdminEventExtension.php#L23

Comment (core23) [reply composer.json]: I mean the `symfony/event-dispatcher` dependency. Do we still need this or can this be replaced with the contract dependency?

Comment (kunicmarko20) [reply composer.json]: contracts just contain interface and event object, so I think we need both

Comment (OskarStark) [reply composer.json]: I think we only need the event symfony/event-dispatcher as it provides the contracts, uses them and have them as dependency. But maybe we need to bump the minimum version

Comment (greg0ire) [reply composer.json]: If we reference the a package in our code, then we must reference it explicitely, and not rely on package A or B to provide it. That rule can be checked with https://github.com/maglnet/ComposerRequireChecker/

Comment (greg0ire) [reply composer.json]: But I think we will need to bump this dependency though: does the 2.8 version class extend a class from the contract package? If not, we have a BC-break.

Comment (phansys) [reply composer.json]: Why you consider it as a BC break? The interface from contracts is not adding new methods, and we aren't using declarations against the contract at all, but checking against `instanceof` to detect if the method signature for the `dispatch()` method must be changed.

Comment (greg0ire) [reply composer.json]: Uuuh right I'm re-reading what I wrote this morning and it no longer makes sense to me :sweat_smile: forget what I said

**Comment by OskarStark** [comment]
If this is ready to merge please restart Travis

**Comment by OskarStark** [comment]
Should we keep this one?

**Comment by phansys** [comment]
I think we should keep this issue, but I'm out of time these weeks.
I'm not able to finish this one in the short time :disappointed:

**Comment by VincentLanglet** [comment]
@phansys Friendly reminder, do you have some time for a rebase ? :)

**Comment by phansys** [comment]
The PR is rebased, but I don't know actually if the change worths. Will `AdminExtractor` keep the translation related methods in the next major?

File: tests/Translator/Extractor/JMSTranslatorBundle/AdminExtractorTest.php [hunk 2]
```diff
@@ -62,8 +63,12 @@ public function setUp(): void
             $this->markTestSkipped('JMS Translator Bundle does not exist');
         }
 
-        $this->fooAdmin = $this->getMockForAbstractClass(AdminInterface::class);
-        $this->barAdmin = $this->getMockForAbstractClass(AdminInterface::class);
+        $this->fooAdmin = $this->getMockBuilder(AbstractAdmin::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (greg0ire) [inline tests/Translator/Extractor/JMSTranslatorBundle/AdminExtractorTest.php]:
```suggestion
        $this->fooAdmin = $this->createMock(AbstractAdmin::class);
```

File: tests/Translator/Extractor/JMSTranslatorBundle/AdminExtractorTest.php [hunk 3]
```diff
@@ -62,8 +63,12 @@ public function setUp(): void
...
+        $this->barAdmin = $this->getMockBuilder(AbstractAdmin::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (greg0ire) [inline tests/Translator/Extractor/JMSTranslatorBundle/AdminExtractorTest.php]:
```suggestion
        $this->barAdmin = $this->createMock(AbstractAdmin::class);
```

File: src/Admin/AbstractAdmin.php [hunk 4]
```diff
@@ -2132,7 +2133,7 @@ public function getTranslationDomain()
      *
      * @deprecated since sonata-project/admin-bundle 3.9, to be removed with 4.0
      */
-    public function setTranslator(TranslatorInterface $translator)
+    public function setTranslator(LegacyTranslatorInterface $translator)
```
Comment (greg0ire) [inline src/Admin/AbstractAdmin.php]: Should this method become deprecated in favor of `setContractTranslator`?

**Comment by greg0ire** [comment]
Tests that use deprecated method should stop doing so or should be marked as deliberately doing so thanks to the `@group legacy` annotation, otherwise the build will keep failing.

File: src/Admin/AbstractAdmin.php (hunk 4 above)
Comment (phansys) [reply src/Admin/AbstractAdmin.php]: I'd prefer not polluting our API for reasons like this. IMO, the better way would be tell the user that in the next major, `setTranslator()` will change the required type for their argument 1.

Comment (phansys) [reply src/Admin/AbstractAdmin.php]: I must back over my own words. `AbstractAdmin::setTranslator()` is deprecated too.

**Comment by phansys** [comment]
Closed in favor of #5932.

**Pull request closed without merging**